packages/libp2p/src/connection-manager/dial-queue.ts

```diff
@@ -230,9 +232,22 @@ export class DialQueue {
         // clean up abort signals/controllers
         signal.clear()
       })
-      .catch(err => {
+      .catch(async err => {
         log.error('dial failed to %s', pendingDial.multiaddrs.map(ma => ma.toString()).join(', '), err)
 
+        if (peerId != null) {
+          // record the last failed dial
+          try {
+            await this.peerStore.patch(peerId, {
```

Comment:
Should this peer also be tagged with a lower value? Let's say the node discovers a new Peer that it has not yet dialled, and this _bad_ peer's threshold has just expired (meaning it's available for a re-dial), shouldn't the newly discovered peer be given a preference in the dial-queue?

Reply:
I agree but also I cannot foresee a scenario where a component would want to prioritize dialling a peer that is currently unreachable, so it should at least be penalized.

Perhaps if there is exists `LAST_DIAL_FAILURE_KEY` in that `peerStore`, it should be sorted below a `peerStore` that doesn't have one (even if the threshold value for the key is expired).

Reply:
> so it should at least be penalized

The problem is the connection manager doesn't know what it means to penalise a peer, because it doesn't know what the value of a tag added by one component means compared to a tag added by another so it cannot alter the tag values without possibly compromising the functioning of the components doing the tagging.

That's why it sets a metadata value to just not dial the peer if it's failed recently.

Reply:
> The problem is the connection manager doesn't know what it means to penalise a peer, because it doesn't know what the value of a tag added by one component means compared to a tag added by another so it cannot alter the tag values without possibly compromising the functioning of the components doing the tagging.

That's why I suggested that we augment the sorting to consider if that key has been added previously so that those peers are de-prioritized.

```typescript
const sortedPeers = shuffledPeers.sort((a, b) => {
      const peerAValue = peerValues.get(a.id) ?? 0
      const peerBValue = peerValues.get(b.id) ?? 0
      
      const peerAhasFailedRecently = a.metadata.has(LAST_DIAL_FAILURE_KEY)
      const  peerBhasFailedRecently = b.metadata.has(LAST_DIAL_FAILURE_KEY)
      
      if (peerAValue > peerBValue && !peerAhasFailedRecently) {
        return -1
      }
      
      if (peerAValue < peerBValue && !peerBhasFailedRecently) {
        return 1
      }
      return 0
})
```

Even after all tag values have been added, it's still _less likely_ that a successful dial would result from a peer that was dialled recently that and failed, than from a newly discovered peer which has not been dialled yet - This assumption is based on the premise that if a peer recently failed to establish a connection, there might be underlying issues causing the failure that haven't been resolved yet (may take longer than 10 minutes) as opposed to when you discover a peer through mechanisms like peer discovery protocols or peer routing, it _should_ indicate that the peer is reachable and active within the network.

Reply:
Ah, right - when you said "Should this peer also be tagged with a lower value?" I thought you meant changing tag values.

The change to the sort function seems reasonable, but I think if you have a small number of peers to choose from (such as during initial startup when you only have bootstrap nodes), you might end up dialling peers you've just failed to dial?  This is expensive and is what this PR is attempting to address.

We might need the opposite though - if we've successfully dialled the peer recently, but we are no longer connected to them, to make sure they bubble up to the top of the autodial queue?

> when you discover a peer through mechanisms like peer discovery protocols or peer routing, it should indicate that the peer is reachable and active within the network.

I don't think this is something we can assume - we might discover peers via a DHT query for example - all that means is they were in the routing tables of another peer, not that they are diallable or even online.

Reply:
> I think if you have a small number of peers to choose from (such as during initial startup when you only have bootstrap nodes), you might end up dialling peers you've just failed to dial? This is expensive and is what this PR is attempting to address.

That's reasonable. I think we can discuss this more on a separate issue.